src/ml/neural_net/combine_base.hpp

```diff
@@ -0,0 +1,253 @@
+ */
+class Completion {
+ public:
+  /** Returns an instance that signals successful completion. */
+  static Completion Finished() { return Completion(); }
+
+  /**
+   * Returns an instance that signals failure, described by the given
+   * exception.
+   */
+  static Completion Failure(std::exception_ptr e) { return Completion(e); }
+
+  bool IsFinished() const { return failure_ == nullptr; }
+
+  /** Returns the exception if a failure and a null pointer otherwise. */
+  std::exception_ptr failure() const { return failure_; }
+
+ private:
+  explicit Completion(std::exception_ptr e = nullptr) : failure_(e) {}
+
+  std::exception_ptr failure_;
+};
+
+/**
+ * Interface for objects that consume values from a Publisher.
+ *
+ * Unless otherwise specified by the concrete implementation, external
+ * synchronization must be used to avoid concurrent calls the Subscriber
+ * interface from different threads.
+ */
+template <typename T>
+class Subscriber {
+ public:
+  /** The type of the values that this Subscriber consumes. */
+  using Input = T;
+
+  virtual ~Subscriber() = default;
+
+  /**
+   * The first signal that a Subscriber receives from a Publisher, passing the
+   * Subscription that the Subscriber can use to control the flow of values.
+   *
+   * A Subscriber may only have one Publisher. If it somehow receives more than
+   * one Subscription, it should call Subscription::Cancel() on any instances
+   * received after the first.
+   *
+   * A Subscriber is explictly allowed to demand values synchronously from
+   * within its implementation of this method.
+   */
+  virtual void Receive(std::shared_ptr<Subscription> subscription) = 0;
+
+  /**
+   * Transmits a value from the Publisher to this Subscriber.
+   *
+   * A Subcriber should never receive more calls to this method than the total
+   * Demand it has requested from its publisher. Subscribers should only demand
+   * more elements from within this method via its return value.
+   */
+  virtual Demand Receive(Input element) = 0;
+
+  /**
+   * Signals completion of the stream of values from the Publisher.
+   *
+   * A Subscriber should not receive any further signals of any kind after
+   * receiving a Completion.
+   */
+  virtual void Receive(Completion completion) = 0;
+};
+
+/**
+ * Interface for objects that produce values on demand from its Subscribers.
+ *
+ * Unless otherwise specified by the concrete implementation, external
+ * synchronization must be used to avoid concurrent calls on multiple threads to
+ * a Publisher, including via the Subscriptions that it passes to its
+ * Subscribers.
+ *
+ * Each concrete implementation defines whether it is unicast or multicast:
+ * whether multiple Subscribers observe the same values or not. (An
+ * implementation might only support one Subscriber, by passing an immediate
+ * Completion to each Subscriber after the first.)
+ */
+template <typename T>
+class Publisher : public std::enable_shared_from_this<Publisher<T>> {
```

Comment:
I think it's better to design that a publisher can also be a subscriber, in which way you can chain publishers.

Reply:
Nothing stops an implementation from deriving from both interfaces

Reply:
But the interface is deceiving, all called `Receive`. The only differences are their parameter lists.

Besides that, Published is being `Subscribed` by subscribers? I feel kind of strange to call `publisher.subcribte(subscriber)`. Shouldn't it be `publisher.distribute(subscriber)` or `subscriber.subcribte(publisher)`?

Reply:
In practice, most users don't call `Receive` explicitly. It's an implementation detail of the class authors. The API is designed to facilitate chaining together operations between a Publisher and a Subscriber. For example, one draft of my client integration had this line of code:
```
data_iterator->AsPublisher()->Map(augmenter)->Map(encoder)->Map(trainer)->Subscribe(sink);
```
The alternative, following your suggestion, would be:
```
sink->Subscribe(data_iterator->AsPublisher()->Map(augmenter)->Map(encoder)->Map(trainer)
```
or maybe something like
```
sink->Subscribe(encoder->Map(augmenter->Map(ToPublisher(data_iterator))))
```

Reply:
Yeah, it's more like a functional programming style. Personally I like it. Or,

```
data_iterator->AsPublisher()->Map(augmenter)->Map(encoder)->Map(trainer)->PublishTo(sink);
```

Anyway, I understand this now.